### PR TITLE
Game Information API

### LINKED ISSUE #152: Learning Flask API and Git
### Problem

In our weekly meeting we have decided to use Flask in our practice app. Thus we have to learn the usage of Flask framework. 

### Solution

I will follow these tutorials and documentations for learning flask api: 
- https://flask.palletsprojects.com/en/2.3.x/quickstart
- https://www.youtube.com/watch?v=GMppyAPbLYk

### Documentation

_No response_

### Additional notes

_No response_

### Reviewers

_No response_

### Deadline

30.04.2023

### PR DESCRIPTION
**Pull request by erkamkavak**
An API which can pull game information from Steam API is added. 

Closes #152 and #160

### CODE REVIEW

File: practice-app/app/views.py [hunk 1]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
     else:
         return render_template("worldtime.html")
 
+def get_game_id_from_name(game_name):
+    steam_api_game_list_url = "https://api.steampowered.com/ISteamApps/GetAppList/v2/"
+    response = requests.get(steam_api_game_list_url).json()
+    all_games_list = response["applist"]["apps"]
```
Comment (omertalib42) [inline practice-app/app/views.py:66]: What happens when the response code is not 200? Handle errors here (or above), based on the status code.

File: practice-app/app/views.py [hunk 2]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
...
+    game_name = "".join(game_name.lower().split())
+    for game in all_games_list:
+        if game_name == "".join(game["name"].lower().split()): 
+            return game["appid"]
+    return -1
```
Comment (omertalib42) [inline practice-app/app/views.py:71]: I'm not sure if this is the best way to indicate there's been a problem as it's way too manual.
Returning None would be better in terms of code readability.

File: practice-app/app/views.py [hunk 3]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
...
 
+def get_game_information_from_id(game_id): 
+    steam_api_get_game_url = "https://store.steampowered.com/api/appdetails?appids={game_id}"
+    response = requests.get(steam_api_get_game_url.format(game_id=game_id)).json()
```
Comment (omertalib42) [inline practice-app/app/views.py:75]: Handle errors by response status code

File: practice-app/app/views.py [hunk 4]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
...
+    game_information = response[str(game_id)]["data"]
+    return game_information
+
+def get_game_information_from_name(game_name): 
+    game_id = get_game_id_from_name(game_name)
+    if game_id == -1:
```
Comment (omertalib42) [inline practice-app/app/views.py:81]: Fix here w/ modifying other function.

File: practice-app/app/views.py [hunk 5]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
...
+    if request.method == "POST":
+        game_name = request.form.get("game_name")
+        if not game_name:
+            error = 'Game name is required!'
+        else:
+            game_information = get_game_information_from_name(game_name)
+            if game_information == None: 
+                error = "Game is not found!"
+            else: 
+                # print(json.dumps(game_information, indent=4))
```
Comment (omertalib42) [inline practice-app/app/views.py:98]: Remove comment

File: practice-app/app/views.py [hunk 6]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
...
+                return render_template('show_game_information.html', response=game_information) 
+    return render_template("get_game_information.html", error=error)
```
Comment (omertalib42) [inline practice-app/app/views.py:100]: Looking at here, having a GET request automatically returns an error... is this intended? If yes, no problem.

File: practice-app/app/views.py [hunk 7]
```diff
@@ -59,4 +60,41 @@ def WorldTime():
...
+
+def get_game_information_from_name(game_name): 
+    game_id = get_game_id_from_name(game_name)
+    if game_id == -1: 
+        return None
+    game_information = get_game_information_from_id(game_id)
+    return game_information
+
+@app.route("/get_game_information", methods= ["GET", "POST"])
+def get_game_information():
```
Comment (omertalib42) [inline practice-app/app/views.py:87]: How about moving all relevant functions to a separate file for better structure? As there will be many more implementations on our codebase, safe to say seperating them will be a better choice.

**Review by omertalib42 — CHANGES_REQUESTED** [review]

File: practice-app/app/views.py (hunk 7 above)
Comment (erkamkavak) [reply practice-app/app/views.py:87]: Should the new seperated file be used from everyone or should it be special to the game information API implementations? Also should I also move the @app.route function too?

File: practice-app/app/views.py (hunk 6 above)
Comment (erkamkavak) [reply practice-app/app/views.py:100]: Yes, it is intended. For the GET request, the error returns as None and it is not shown in the html template. For the POST request, if there is an error it shows the error in the html.

**Comment by erkamkavak** [comment]
Since we need to add database and POST request to our APIs, I am converting this PR to draft. After adding databases, I will reopen the PR.

File: practice-app/app/views.py (hunk 3 above)
Comment (omertalib42) [reply practice-app/app/views.py]: Make this f-string.

File: practice-app/app/game_information_api.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+from flask import render_template, request, redirect
+from sqlalchemy import Column, String, Integer, ForeignKey, PrimaryKeyConstraint
+from sqlalchemy.orm import Mapped, mapped_column
+from flask_login import login_required, current_user
+
+from .views import app, Base, session
+
+class FavoriteGame(Base):
+    __tablename__ = 'FavoriteGameTable'
+    id = Column(Integer, autoincrement=True)
```
Comment (omertalib42) [inline practice-app/app/game_information_api.py:11]: id fields are null in the db, ~~fix required~~ what do you think of it?

File: practice-app/app/game_information_api.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+@app.route("/show_all_favorites", methods=["POST"])
+@login_required
+def show_all_favorites():
+    all_user_games = FavoriteGame.query.filter_by(user_id=current_user.id).all()
+    return render_template("show_all_games.html", user_games=all_user_games)
+
+@app.route("/add_to_favorites", methods=["POST"])
+@login_required
+def add_game_to_favorites(): 
+    if request.method == "POST":
```
Comment (omertalib42) [inline practice-app/app/game_information_api.py:69]: GET requests are automatically handled with 405 code, so no need to check it here.

File: practice-app/app/game_information_api.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+@login_required
+def add_game_to_favorites(): 
+    if request.method == "POST": 
+        game_name = request.form.get("game_name")
+        game_id = int(request.form.get("game_id"))
+        log = None
+        if not game_name or not game_id: 
+            log = "Problems occurred! Game could not be added to favorites."
+        else: 
+            previously_added = FavoriteGame.query.filter_by(user_id=current_user.id, game_id=game_id).all()
```
Comment (omertalib42) [inline practice-app/app/game_information_api.py:76]: A fun test for you: Try adding & reloading the game "Smite". 
What do you see?

File: practice-app/app/game_information_api.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+            else: 
+                favorite_game = FavoriteGame(
+                    user_id=current_user.id,
+                    game_id=game_id,
+                    game_name=game_name
+                )
+                session.add(favorite_game)
+                session.commit()
+                log = "Game added to favorites!"
+        print(log)
```
Comment (omertalib42) [inline practice-app/app/game_information_api.py:88]: Remove print

File: practice-app/app/game_information_api.py (hunk 8 above)
Comment (omertalib42) [reply practice-app/app/game_information_api.py:11]: image

File: practice-app/app/game_information_api.py (hunk 11 above)
Comment (omertalib42) [reply practice-app/app/game_information_api.py:88]: I -personally- do not prefer to merge to main with unnecesary comments & prints. 
If you need them, using [logging](https://realpython.com/python-logging/) would be a better practice imo.

File: practice-app/app/game_information_api.py
Comment (omertalib42) [inline practice-app/app/game_information_api.py:1]: You may also consider using [black](https://black.readthedocs.io/en/stable/), as mentioned in the meeting. 

I'm not enforcing it of course, just a suggestion.

**Review by omertalib42 — CHANGES_REQUESTED** [review]

**Comment by omertalib42** [comment]
It's good in general, just some cleanups.

File: practice-app/app/game_information_api.py (hunk 10 above)
Comment (omertalib42) [reply practice-app/app/game_information_api.py:76]: We may talk about the ways of fixing this, if you want.

File: practice-app/app/game_information_api.py (hunk 8 above)
Comment (erkamkavak) [reply practice-app/app/game_information_api.py]: Fixed now.

File: practice-app/app/game_information_api.py (hunk 10 above)
Comment (erkamkavak) [reply practice-app/app/game_information_api.py]: Nice catch! I guess it is fixed now.

**Pull request merged**